### PR TITLE
Userlist lambda

### PR DESCRIPTION
**Pull request by josephgarner**
Few things in the PR

- User list end point created, this however is a deprecated feature and may be removed after a discussion
- Updated endpoints calling Hack details to append user information to the team so the frontend does not have to sort the data

Latest Update: Changed Hackdetails.js and GetHacks.js files replacing the for loops

### CODE REVIEW

File: getUsers.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
+const mongoose = require("mongoose");
```
Comment (jpeyper) [inline getUsers.js]: is this file still required at all?

File: getHacks.js [hunk 2]
```diff
@@ -18,16 +18,29 @@ export const list = async (event, context) => {
...
+  const Teams = conn.model("User");
+
   /**
    * To be used in the future This will return One document
    * const doc = await Query.findOne({_id: '5e6094446a56971ad6a32d7b'});
    */
   try {
     const doc = await Query.find();
+    const users = await Teams.find();
+    for(const hack_key in doc){
```
Comment (jpeyper) [inline getHacks.js]: I feel like these loops are reimplemented a fair bit... can we have a `getUser` lambda that we invoke (in parallel?) to fill out these details in a common way? or extract it into another function that is imported into all the places.

I'm not sure what best practice is around lambdas invoking other lambdas...

File: getUsers.js (hunk 1 above)
Comment (josephgarner) [reply getUsers.js]: Unsure, I will need to discuss with Stanley

Comment (HuangStanley050) [reply getUsers.js]: yeah it is required, we still need the mongoose to make connection to the database

Comment (jpeyper) [reply getUsers.js]: sorry, i meant the whole lamdba, not just this line (does the front end still get all the users at any point)

File: getHacks.js (hunk 2 above)
Comment (josephgarner) [reply getHacks.js]: I'll look into it

Comment (josephgarner) [reply getHacks.js]: Looking further into this, its strongly suggested against calling lambdas from other lambdas, its possible and every comment I found says its "easy"... Anyway replaced those loops with a function from mongoose.js called populate, a simple version of an SQL Join pretty much.

File: getUsers.js (hunk 1 above)
Comment (josephgarner) [reply getUsers.js]: File Has been removed for now

File: getHacks.js [hunk 3]
```diff
@@ -18,16 +18,13 @@ export const list = async (event, context) => {
...
+    conn.model("User", new mongoose.Schema({ name: String }));
   }
   const Query = conn.model("Hack");
-  /**
-   * To be used in the future This will return One document
-   * const doc = await Query.findOne({_id: '5e6094446a56971ad6a32d7b'});
-   */
   try {
-    const doc = await Query.find();
+    const doc = await Query.find().populate('team', '-email', 'User');
```
Comment (jpeyper) [inline getHacks.js:27]: I thought mongoose would only populate what was in the schema, and you wouldnt need to redact the email field.

File: joinHack.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  if (conn == null) {
+    conn = await mongoose.createConnection(uri, {
+      bufferCommands: false,
+      bufferMaxEntries: 0
+    });
+    conn.model(
+      "Hack",
+      new mongoose.Schema({team: Array})
+    );
+    conn.model("User", new mongoose.Schema({ name: String, email: String }));
```
Comment (jpeyper) [inline joinHack.js]: missed one

File: package.json [hunk 5]
```diff
@@ -15,10 +15,11 @@
...
     "aws-sdk": "^2.636.0",
     "serverless-bundle": "^1.2.5",
-    "serverless-dotenv-plugin": "^2.1.1",
+
     "serverless-offline": "^5.3.3"
   },
   "dependencies": {
-    "mongoose": "^5.9.3"
+    "mongoose": "^5.9.3",
+    "serverless-dotenv-plugin": "^2.1.1"
```
Comment (jpeyper) [inline package.json:23]: is it? I would have though its all done at deployment time and not at runtime (therefore still a devdep)

**Review by jpeyper — CHANGES_REQUESTED** [review]

File: getHacks.js (hunk 3 above)
Comment (josephgarner) [reply getHacks.js:27]: I had the same thought, however it still grabs the email

File: package.json (hunk 5 above)
Comment (HuangStanley050) [reply package.json:23]: "serverless-dotenv-plugin": "^2.1.1" 

I added that to dependencies is because I want to see if it will pack in the .env.prod file I set up in github action because during local deploy apparently this plugin take the .env file and inject all envs for the lambdas.

File: joinHack.js (hunk 4 above)
Comment (josephgarner) [reply joinHack.js]: Yes I did, fixed now

File: package.json (hunk 5 above)
Comment (HuangStanley050) [reply package.json:23]: maybe I also need to add the serverless-webpack plugin to the dependencies to compact the functions but can be implemented later, right now just want to see if this works in github actions.

Comment (jpeyper) [reply package.json:23]: all those actions should be happening at the bundling and deployment time though, the (which devDependencies will be present for). the dependencies are only required for stuff that in used during runtime (like mongoose).

Basically if you dont import/require it, its a devDependency.

At the end of the day, given we are webpacking and serverless is taking care of the deployment, its likely that it doesnt matter and everything could be either a dependency or a devDependency and everything will still work.

**Review by jpeyper — APPROVED** [review]
